### DISCUSSION TITLE
tRPC v10 DX related notes

### DISCUSSION THREAD

**Discussion by jlalmes** [trpc/trpc#2164] [General]
## tRPC v10 DX related notes.

Some initial thoughts as I scan through the `next` branch.

---

- Should we block double `input/output/meta` calls?
```ts
t.procedure
  .input(z.string())
  .input(z.number()) // 👈  TS error here?
```

- Should we block `subscription` once `output` is added?
```ts
t.procedure
  .output(z.string())
  .subscription(/* ... */) // 👈  TS error here?
```

- Should we supply `observable` to `subscription` procedures instead of importing it?
```diff
- import { observable } from '@trpc/server/observable';

t.procedure
-  .subscription(({ ctx, input } =>
+  .subscription(({ ctx, input }, observable /* 👈 ? */) =>
```

- Should `subscription` typing require an `observable` as return type. Like v9 requires `Subscription` return type.
```ts
t.procedure
  .subscription(() => {
    return null; // 👈  TS error here?
  })
```

- Should `createTRPCClientProxy` be able to optionally skip `createTRPCClient` step?
```diff
- const client = createTRPCClient<AppRouter>({ url })
- const proxy = createTRPCClientProxy(client);
+ const client = createTRPCClientProxy<AppRouter>({ url }); // 👈 ?
```

- Likewise, should `createReactQueryHooksProxy` be able to optionally skip `createReactQueryHooks` step?
```diff
- const hooks = createReactQueryHooks<AppRouter>()
- const trpc = createReactQueryHooksProxy(hooks);
+ const trpc = createReactQueryHooksProxy<AppRouter>(); // 👈 ?
```

- Additionally to above two points, should we *deprecate* `createTRPCClient` & `createReactQueryHooks`?

**Comment by jlalmes** [comment]
> Should we block `subscription` once `output` is added?

Alternatively we could allow `output`, then infer (& validate) the `next(value)` type from the subscription's observable.

```ts
t.procedure
  .output(z.string())
  .subscription(() => {
    return observable((observer) => {
      observer.next('jlalmes') // 👌 valid - next expects `string`
      observer.next(1234) // 👈 TS error
    })
  })
```

**Comment by KATT** [comment]
Thank you!


> * Should we block double `input/output/meta` calls?
> 
> ```ts
> t.procedure
>   .input(z.string())
>   .input(z.number()) // 👈  TS error here?
> ```

Ideally, yes. It already fails runtime, but would be a nice DX change to fail at compile time. Feel free to make an issue 


> * Should we block `subscription` once `output` is added?
> 
> ```ts
> t.procedure
>   .output(z.string())
>   .subscription(/* ... */) // 👈  TS error here?
> ```

Suuuure...... or maybe the output the subscription emits could be parsed?

> * Should we supply `observable` to `subscription` procedures instead of importing it?
> 
> ```diff
> - import { observable } from '@trpc/server/observable';
> 
> t.procedure
> -  .subscription(({ ctx, input } =>
> +  .subscription(({ ctx, input }, observable /* 👈 ? */) =>
> ```


This is nice, I would although not have it as it's own arg _(might've been a typo)_:


```diff
t.procedure
-  .subscription(({ ctx, input }, observable /* 👈 ? */) =>
+  .subscription(({ ctx, input, observable /* 👈 ? */ }) =>
```

> * Should `subscription` typing require an `observable` as return type. Like v9 requires `Subscription` return type.
> 
> ```ts
> t.procedure
>   .subscription(() => {
>     return null; // 👈  TS error here?
>   })
> ```

Yes.

> * Should `createTRPCClientProxy` be able to optionally skip `createTRPCClient` step?
> 
> ```diff
> - const client = createTRPCClient<AppRouter>({ url })
> - const proxy = createTRPCClientProxy(client);
> + const client = createTRPCClientProxy<AppRouter>({ url }); // 👈 ?
> ```
> * Likewise, should `createReactQueryHooksProxy` be able to optionally skip `createReactQueryHooks` step?
> 
> ```diff
> - const hooks = createReactQueryHooks<AppRouter>()
> - const trpc = createReactQueryHooksProxy(hooks);
> + const trpc = createReactQueryHooksProxy<AppRouter>(); // 👈 ?
> ```
> 
> * Additionally to above two points, should we _deprecate_ `createTRPCClient` & `createReactQueryHooks`?

I'm down for these changes, but I'm unsure about a good migration path in all cases. Feel free to open a discussion on this topic alone or chip in here where we discuss something very related: https://github.com/trpc/trpc/pull/2119#issuecomment-1179537368

---


Neither of these seems critical for alpha/beta releases, but feel free to create issues for all of them. @sachinraja @mmkal, anything you see here that should be added/amended?

**Comment by sachinraja** [comment]
I agree with all of @KATT's notes.

> Should we block double input/output/meta calls?

IMO, yes. However, it's not very important to me either so I don't think it's worth prioritizing right now.

> Should subscription typing require an observable as return type.

Yes. I think it's better to be more strict here.

Great suggestions @jlalmes!

**Comment by jlalmes** [comment]
What if `createTRPCClient` supported both `v9` & `v10`?

```diff
- const client = createTRPCClient<AppRouter>({ url });
- const proxy = createTRPCClientProxy(client);

- client.query('ping');
- proxy.ping.query();

+ const client = createTRPCClient<AppRouter>({ url });

+ client.query('ping'); // 👈 strikethrough (.query is @deprecated)
+ client.ping.query();
```

**Reply by KATT** [reply]
I don't _think_ I want to mix them. I had it like that first but then you expose a lot of stuff that will never be used in your actual usage.

There's also the added complexity that _if_ someone wants to [support for instance IE11](https://caniuse.com/?search=Proxy), you might not want to use the `Proxy` client at all.

Unsure what the right balance is between:

- Ease of **migration path** - I don't think I want to _force_ ppl to change anything on the client
- Ease of **boilerplate** - I understand it'd be nice to only have exactly 1 setup thing for the client, but it's also a one-time thing to do, so I'm a bit torn.
- Ease of **configuration** - maybe someone doesn't want to use `Proxy`
- Ease of **use** - I don't want to unnecessarily leak the "old way" of doing  things unless they explicitly want that b/c they don't want to use `Proxy`

**Reply by jlalmes** [reply]
Good argument - let's leave as-is.

**Reply by mmkal** [reply]
Copying a comment from #2217 which I think toes the line between those concerns pretty well:

how about a single create client function like:

```ts
import {createClient} from '@trpc/client'

const client = createClient<AppRouter>({url: ..., fetch: ...})

await client.query('foo', {abc: 123})
await client.mutation('bar', {xyz: 456})

await client.proxy.foo.query({abc: 123})
```

Projest that want to always use the proxy API could make a module like:

```ts
import {createClient} from '@trpc/client'

export const {proxy: trpc} = createClient({url: ..., fetch: ...})
```

Then anyone needing to use the client would just do:

```ts
import {trpc} from './utils/trpc'

await trpc.foo.query({abc: 123})
```

If worried about protecting people using the Proxy API from baggage or something that fails in IE, the `.proxy` could be lazy, ie `get proxy() { return new Proxy(...) }` or similar.

**Reply by KATT** [reply]
We could also provide a new way of defining the client and only have the current `createTRPCClient` for interop & ppl who for some reason might not want to use `Proxy`?

**Comment by aradalvand** [comment]
If you decide [not to smurf](https://github.com/trpc/trpc/discussions/2131), one option could be to depracate `createTRPCClient` and introduce a new `createClient` function which is the new `Proxy` version.

And also please for the love of god, [please don't use the `proxy.getSomething.query()` syntax](https://github.com/trpc/trpc/issues/2217)...
